### ISSUE TITLE
Merge DynamicSupervisor in

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#5267]
This is a reminder that, once DynamicSupevisor is merged, we will need to implement the code that validates typespecs and ensure both Supervisor and DynamicSupervisor return the same value on `init` (since we can't change what `Supervisor.init` returns). This is scheduled for v1.5.

**Comment by michalmuskala** [comment]
Would inclusion of DynamicSupervisor mean soft-deprecation (and maybe a full one eventually) of simple_one_for_one strategy in regular supervisors?

**Comment by josevalim** [comment]
We would likely never remove it, because we are simply forwarding it to the Erlang supervisor, but yes.

**Comment by fishcakez** [comment]
DynamicSupervisor only supports temporary children btw

**Comment by fishcakez** [comment]
> DynamicSupervisor only supports temporary children btw

This is not true it supports all restart values.

**Comment by AndrewDryga** [comment]
Hey @josevalim. Can you give some additional details what should be done?

1. DynamicSupervisor was renamed
2. Does it mean that this task requires simple copy-paste merge (and, possibly, rename) of [this module](https://github.com/elixir-lang/gen_stage/blob/2f8cf4d1923c1cd35cd54d7630dd87a659031954/lib/consumer_supervisor.ex) and it's tests into elixir repo?

**Comment by josevalim** [comment]
This one is currently assigned to me which means I am working on it. ;) But the idea is that we don't want the GenStage features of DynamicSupervisor/ConsumerSupevisor and we need to keep Supervisor compatibility, so it is a subset of what is in ConsumerSupervisor today.

**Comment by AndrewDryga** [comment]
Ah, ok. I'll try to find another monster to beat :).

**Comment by josevalim** [comment]
We will first propose to include this as part of OTP. Therefore it won't be part of 1.5.

**Comment by josevalim** [comment]
This will be added to v1.6. v1.6 requires OTP 19, which is the minimum OTP version that supports custom supervisors at the application root.

**Comment by myronmarston** [comment]
I've always found `:simple_one_for_one` to be confusing, so I'm looking forward to seeing what Elixir 1.6 will have here.  I did want to ask one question about the planned API for this, because there's an aspect of `:simple_one_for_one` that I've always found a bit frustrating, and the new `DynamicSupervisor` seems like the right time to address it.

Specifically, the way that `:simple_one_for_one` combines arguments given in the original child spec and arguments provided to `Supervisor.start_child` has confused me and I always have to look it up in the docs every time I deal with it.  And usually I wind up defining a special `start_link` function (e.g. `start_link/2` instead of `start_link/1`) specifically for this case, which is only ever used when starting a GenServer via its supervisor (whereas my tests tend to use `start_link/1` with a single options KW list).

Instead of Elixir defining how the arguments given to the child spec are combined with the args given to `Supervisor.start_child`, I'd like a way for the developer to define it when they define their child spec.  I'm thinking something like:

``` elixir
defmodule MyDynamicSupervisor do
  def start_link(options) do
    foo = Keyword.fetch!(options, :foo)
    child_template = {ChildTemplate, &Keyword.put(&1 :foo, foo)}
    Supervisor.start_link([child_template], strategy: :dynamic_one_for_one)
  end
end
```

I don't know what the planned API for dynamic supervisors are, so I have no idea if this direction would work (or if `:dynamic_one_for_one` is the right strategy for this case), but I'd like to be able to provide a function that combines the arguments provided to the supervisor with the arguments provided for the specific child in whatever way makes sense for my application.  It's also a lot more explicit than the current way, where the arg lists are concatenated into a longer arg list, and you just have to know from the docs that that will happen.

Can an improvement like this be included in the new `DynamicSupervisor` API?

If there's a doc that outlines the planned API and already answers this, feel free to point me to it :).

Thanks!

**Comment by josevalim** [comment]
@myronmarston this is something we have to address because the new child_specs make it allegedly worse but we haven't put much thought into it yet.

By having its own modules, we can be sure to provide at least better docs though.

**Comment by josevalim** [comment]
Closing in favor of #7006.
